libursa/QueryGraph.cpp

```diff
@@ -29,40 +32,116 @@ QueryGraph QueryGraph::dual() const {
+template <typename T, T TM()>
 class NodeState {
     std::vector<NodeId> ready_predecessors_;
     uint32_t total_predecessors_;
-    QueryResult state_;
+    T state_;
 
    public:
-    NodeState() : state_(QueryResult::everything()), total_predecessors_(0) {}
+    // Constructs a new instance of NodeState with provided constructor.
+    NodeState() : state_(TM()), total_predecessors_(0) {}
 
-    const QueryResult &state() const { return state_; }
+    // Gets a reference to internal state.
+    const T &get() const { return state_; }
 
+    // Moves the provided value into internal state.
+    void set(T &&state) { state_ = std::move(state); }
+
+    // Gets a vector of already processed predecessors.
     const std::vector<NodeId> &ready_predecessors() const {
         return ready_predecessors_;
     }
 
-    void set(QueryResult &&state) { state_ = std::move(state); }
-
+    // Used for counting incoming edges into the graphs node.
     void add_predecessor() { total_predecessors_++; }
 
+    // Called when a predecessor was processed and becomes ready.
     void add_ready_predecessor(NodeId ready) {
         ready_predecessors_.push_back(ready);
     }
 
+    // Node is assumed ready when all of its predecessors were processed.
     bool ready() const {
         return ready_predecessors_.size() >= total_predecessors_;
     }
 };
 
-QueryResult masked_or(const std::vector<const QueryResult *> &to_or,
+// This class can be used to traverse QueryGraph DAG in topological order.
+// It's guaranteed that during processing of every node, all it's parents
+// were already processed.
+template <typename T, T TM()>
+class InorderGraphVisitor {
+    std::vector<NodeId> ready_;
+    const std::vector<QueryGraphNode> *nodes_;
+    std::vector<NodeState<T, TM>> state_;
+
+   public:
+    // Creates a new instance of InorderGraphVisitor. `ready` vector should
+    // contain verticles with no incoming edges (aka graph sources).
+    InorderGraphVisitor(std::vector<NodeId> ready,
+                        const std::vector<QueryGraphNode> *nodes)
+        : ready_(std::move(ready)), nodes_(nodes), state_() {
+        state_.reserve(nodes_->size());
+        for (size_t i = 0; i < nodes_->size(); i++) {
+            state_.push_back(NodeState<T, TM>());
+        }
+        for (size_t ndx = 0; ndx < nodes_->size(); ndx++) {
+            for (NodeId target : (*nodes)[ndx].edges()) {
+                state_.at(target.get()).add_predecessor();
+            }
+        }
+    }
+
+    // True if traversal was finished and visit queue is empty.
+    bool empty() const { return ready_.empty(); }
+
+    // Returns id of next graph's node to process.
+    NodeId next() {
+        NodeId nextid = ready_.back();
+        ready_.pop_back();
+        for (const auto &succ : (*nodes_)[nextid.get()].edges()) {
+            state_[succ.get()].add_ready_predecessor(nextid);
+            if (state_[succ.get()].ready()) {
```

Comment:
extract state_[succ.get()] into new variable?

Reply:
not sure if worth it, but done :thi

Reply:
done